[PyTorch][Edge] Add api to get bytecode model version

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56801 [PyTorch][Edge] Add api to get bytecode model version**

Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

Commit:
[PyTorch][Edge] Add api to get bytecode model version

Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

raziel commented:
why does this PR depend on backport already?

cccclai commented:
> why does this PR depend on backport already?

Hmm I think backport depend on this pr? Backport needs to know the model version.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"




Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

dhruvbird commented:
I'm assuming that there's a cheaper way to get the bytecode version number than reading the entire bytecode.pkl along with the associated tensors?

JacobSzwejbka commented:
> I'm assuming that there's a cheaper way to get the bytecode version number than reading the entire bytecode.pkl along with the class resolution (since this requires the torchbind objects to exist)?

I'll have to do a similar thing for the model ops as well. Curious if those apis will ever be called seperately, or if its worth just combining them into one to save a parsing pass over bytecode.pkl

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

cccclai commented:
> I'm assuming that there's a cheaper way to get the bytecode version number than reading the entire bytecode.pkl along with the class resolution (since this requires the torchbind objects to exist)?

Can you elaborate a bit more about it? Given a model, the bytecode version exist in bytecode.pkl and we can read the bytecode.pkl to get the bytecode version number. The api can also be used by other users (like model delivery) to check bytecode compatibility, so it is also exposed to python.

cccclai commented:
> > I'm assuming that there's a cheaper way to get the bytecode version number than reading the entire bytecode.pkl along with the class resolution (since this requires the torchbind objects to exist)?
> 
> I'll have to do a similar thing for the model ops as well. Curious if those apis will ever be called seperately, or if its worth just combining them into one to save a parsing pass over bytecode.pkl

There will be some cases to be called separately. When backporting a model (most likely the bytecode schema change) from a new version to old version, there is no need to check operator schema. A very rough example case will be:

```
model_bytecode_version = _get_model_bytecode_version(model)
runtime_bytecode_version = retrive_from_client() -> # basically client call _get_runtime_bytecode_version() and send to server
if(bytecode_version > runtime_bytecode_version):
    backport_success = backport(model, runtime_bytecode_version)
```
In this case, we don't need to know operator shema.

In terms of reading `bytecode.pkl` twice, that's a good point! I can add a function in cpp to get_model_bytecode_version(bytecode_vals), which accept a parsed bytecode_vals. In this case, we just need to call `readArchive` once and pass the parsed vals to the function. Something like this:

```
bvals = readArchive("bytecode")
bytecode_version = get_bytecode_version(bvals)
ops = get_ops_schema(bvals)
```

what do you think?

raziel commented on torch/csrc/jit/mobile/model_bytecode_version.cpp:
```diff
@@ -0,0 +1,175 @@
+  }
+}
+
+c10::IValue readArchive(
```
Are all these utilities copy pasted from somewhere else?

cccclai replied:
Currently `mobile/import.cpp`, `jit/serialization/import.cpp` and this file uses the generic function `torch::jit::readArchiveAndTensors(archive_name, type_resolver, obj_loader, device, stream_reader)` to read archive, after #57098 lands. 

The `type_resolver` and `obj_loader` lambda functions here are the same as `mobile/import.cpp`, but different from `jit/serialization/import.cpp`. I can probably try separating them out from `mobile/import.cpp` and reuse it here.

raziel commented on torch/csrc/jit/serialization/import_read.cpp:
```diff
@@ -29,10 +29,33 @@ IValue readArchiveAndTensors(
-    return std::get<0>(stream_reader.getRecord(ss));
+    std::size_t found = name.find(slash);
+    std::stringstream ss;
+    // In bytecode version 4, the tensor root_key doesn't include the parent
```
Is this logic here because we don't want to require a backport to v4?
don't we anyway will need backport logic once we bump to v5 since there's other changes beyond the / ?

cccclai replied:
Oh actually this logic belongs to #56002 (reuse constant table), because it's not related to get bytecode model version. It's a bit messed up when I restack the changes. This is for backward compatibility. It means that client running v5 can load model v4.

JacobSzwejbka commented on torch/csrc/jit/mobile/model_bytecode_version.h:
```diff
@@ -0,0 +1,26 @@
+namespace jit {
+namespace mobile {
+
+// The family of methods below to get bytecode version from a model
```
Does bytecode exist outside of the mobile space?

iseeyuan replied:
For serialization APIs (both export and import) we don't put them in mobile namespace. 
1. Those APIs are also used in server side. For example, at deployment stage.
2. torch::jit::_load_for_mobile, the name may already differentiates from torch::jit::load. Adding another namespace is not necessary.

JacobSzwejbka replied:
Theres nothing about the name of this method that screams mobile to me though, and the the model ops stuff will have the same problem but worse. I'm not sure I see the harm in essentially labeling that hey this stuff is mobile specific it wont work unless you pass in the file path of a mobile module.

iseeyuan replied:
@JacobSzwejbka Either function name or file path could give a hint.

iseeyuan replied:
1. Namespace: I think it's good to remove mobile in this PR as discussed above.
2. File organization: it can be done in a separate PR later (tracking in T89985123) 
* The serialization related codes (`import.h/cpp`, export_data.cpp`, compatibility.h/cpp`, etc) should be moved to torch/csrc/jit/serialization/mobile. 
* torch/csrc/jit/mobile should be renamed to torch/csrc/jit/mobile_runtime to make this folder runtime specific.

JacobSzwejbka commented:
@cccclai to verify you'll keep the python bindings in this diff right? I think it'll be a useful debugging tool at the very least for the model stuff to have a python api

iseeyuan commented on torch/csrc/jit/mobile/model_bytecode_version.cpp:
```diff
@@ -0,0 +1,175 @@
+  };
+
+  auto obj_loader = [&](at::StrongTypePtr type, IValue input) {
+    auto cls = type.type_->expect<at::ClassType>();
```
We are not interested in special types. Could we just return a dummy IValue?

iseeyuan commented on torch/csrc/jit/mobile/model_bytecode_version.cpp:
```diff
@@ -0,0 +1,175 @@
+
+namespace {
+
+TypePtr resolveTypeName(
```
We should be able to pass a dummy type resolver here, because we are only interested in the bytecode literals, not the classes.

cccclai replied:
Same as readArchive

kimishpatel commented:
Is it possible to do something similar to show_pickle where you dont add all the dependency for class types etc just to read byte code? 
Or maybe something like this? https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/import_read.cpp#L38-L44. Where type resolver, object loader etc, can be null?
On the other hand if this is used later with backporting then it might be ok? Not clear to me what dependencies you need in order to read just the bytecode. If you dont need type resolvers and obj loaders (which I think are in data archive), then you can get rid of those even for backporting.

iseeyuan reviewed (changes requested):
1. Please pass a dummy type resolver and obj loader to the unpickler, instead of a copy from runtime import. Please refer to the inline comments for details.

2. Consider having an API returning the big tuple of bytecode, for all usages  on querying the bytecode literals. Not only for version number, but also for operator table like @JacobSzwejbka suggested, or for any future queries.

iseeyuan commented:
> I'm assuming that there's a cheaper way to get the bytecode version number than reading the entire bytecode.pkl along with the class resolution (since this requires the torchbind objects to exist)?

pickle file is a binary file. There's some interpreter to parse the content, including the version number. We don't need the actual class at this time, so we could pass a dummy class resolver. Please let us know if there is even cheaper ways.

cccclai commented:
> @cccclai to verify you'll keep the python bindings in this diff right? I think it'll be a useful debugging tool at the very least for the model stuff to have a python api

Yes python binding will be kept.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

cccclai commented:
dummy type_resolver doesn't work for some bytecode.pkl in prod models. Paste the model bytecode.pkl in D27961417. 

I suggestion we keep the type_resolver and obj_loader for now, so it's the same as `mobile/import.cpp`, in case the backport process crash in edge case. After the api is deployed and integration test are established, we can safely remove some parts.

kimishpatel commented:
> dummy type_resolver doesn't work for some bytecode.pkl in prod models. Paste the model bytecode.pkl in D27961417.

Do you know why?

cccclai commented:
> > dummy type_resolver doesn't work for some bytecode.pkl in prod models. Paste the model bytecode.pkl in D27961417.
> 
> Do you know why?

I used a nullptr and it fell here https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/unpickler.cpp#L609. @iseeyuan suggests me to use a dummy one instead of nullptr.

kimishpatel commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -0,0 +1,91 @@
+    // files that do not start with the file entry, so it is relatively safe to
+    // perform this check.
+    TORCH_WARN("The zip file might be problematic. Please check it again.");
+    return true;
```
This is a little strange. You want to return false when zip file is not ok.

Also how id 0x80 and 0x02 is derived as values.

dhruvbird commented:
wow! I come back 2 days later and this diff is all cleaned up :)

My main concern was that it seemed like we were copy-pasting a bunch of code just for being able to get the bytecode version out from the model. The new version seems to:
1. Not have much copy-paste
2. Tries to make the `bytecode.pkl` parsing a general utility which "other" (future) features can also use, and isn't tied specifically to the purpose of getting the bytecode version out from the model.

Carry on!

kimishpatel commented:
> dummy

I see that in your latest you are using nullptr for type resolver and obj loader. Does that work now?

> I used a nullptr and it fell here https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/unpickler.cpp#L609. @iseeyuan suggests me to use a dummy one instead of nullptr.

I see that in your latest commit you are using nullptr for type resolver and obj loader. So it works now? I wonder what was the cause of failure before.

cccclai commented:
nullptr doesn't work. I try to use a dummy type_resolver but it doens't work out of box. Something like this: 

```
struct TORCH_API Dummy : public c10::Type {
  static DummyPtr create() {
    return DummyPtr(new Dummy()); // NOLINT(modernize-make-shared)
  }
  bool operator==(const Type& rhs) const override {
    return rhs.kind() == kind();
  }
  std::string str() const override {
    return "Scalar"; // match what PythonArgParser says for clarity
  }
  static const at::TypeKind Kind = at::TypeKind::ListType;
// global singleton
//  static DummyPtr get();

  Dummy(at::TypeKind kind = at::TypeKind::ListType) : Type(kind) {}
 protected:

  std::string annotation_str_impl(at::TypePrinter printer = nullptr) const override {
    return "dummy"; // technically not a valid python type, but
// we need to use it when parsing back in annotations
// for implicit conversions
  }
};

```
and my type_resolver is
```
at::TypePtr resolveTypeName(const c10::QualifiedName& qn) {
    return std::make_shared<Dummy>();
}
auto type_resolver = [&](const c10::QualifiedName& qn) {
   return c10::StrongTypePtr(compilation_unit, resolveTypeName(qn));
 };

```
I was thinking that a dummy handler can be used for customized type, but this dummy type actually falls for a common type:
```
type: List[Optional[Tensor]]
```

I'm thinking that we can at least reuse the one from `mobile/import.cpp` for now, so we can make sure it can handle all edge cases. After integration test is in place, we can replace it with the dummy type_resolver. At this moment, If I managed to have a dummy type_resolver/obj_loader and have it working, I feel like I don't have enough test case to test it. If I miss anything, feel free to let me know

cccclai commented:
> > dummy
> 
> I see that in your latest you are using nullptr for type resolver and obj loader. Does that work now?
> 
> > I used a nullptr and it fell here https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/unpickler.cpp#L609. @iseeyuan suggests me to use a dummy one instead of nullptr.
> 
> I see that in your latest commit you are using nullptr for type resolver and obj loader. So it works now? I wonder what was the cause of failure before.

What failure do you refer to? Using nullptr? unpickler will check if type_resolver is nullptr when none of if statements works,
  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/unpickler.cpp#L609

kimishpatel commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -0,0 +1,91 @@
+  c10::optional<at::Device> device;
+
+  auto ivalues = torch::jit::readArchiveAndTensors(
+      archive_name, nullptr, nullptr, device, stream_reader);
```
I was referring to this. Does this work only in some cases?

cccclai replied:
Yup it only works for in some minimum cases.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

cccclai commented:
Main change:
1. Reuse the `type_resolver` and `obj_loader` from `mobile/import.cpp` in the latest commit.

Note:
Attempt to follow @iseeyuan's suggestion to use a simpler `type_resolver` and `obj_loader` for reading pkl file. Likely it will take more than a week to implement and fully test. At this moment, we don't have tests set up to cover all models. Recommend to reuse the ones from `mobile/import.cpp`, which has been fully tested, and we don't need to worry about failure during model delivery. After integration tests are added, we can switch to the simpler `type_resolver` and `obj_loader`.

raziel reviewed (commented):
Overall this seems cleaner (less copy paste) but I'll like @iseeyuan and @kimishpatel to complete the review since they're more familiar with the code.

I still suggest creating a task for tracking this:

"Attempt to follow @iseeyuan's suggestion to use a simpler type_resolver and obj_loader for reading pkl file in pr #57453. Likely it will take more than a week to implement and fully test. At this moment, we don't have tests set up to cover all models. Recommend to reuse the ones from mobile/import.cpp, which has been fully tested, and we don't need to worry about failure during model delivery. After integration tests are added, we can switch to the simpler type_resolver and obj_loader."

I agree not making it a blocker but shouldn't also be left unimplemented for long.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]

cccclai commented:
Add the task number in the latest change.

cccclai commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -0,0 +1,107 @@
+  std::shared_ptr<CompilationUnit> compilation_unit =
+      std::make_shared<CompilationUnit>();
+
+  // TODO (T90180710): Simplify type_resolver and obj_loader when getting
```
Here is the todo and task number

iseeyuan commented on torch/csrc/jit/mobile/import.h:
```diff
@@ -72,6 +72,20 @@ void _load_extra_only_for_mobile(
+// Currently used by both mobile/import.cpp and model_compatibility.cpp.
+// Should be removed after model_compatibility.cpp start using simplified
+// version type_resolver and obj_loader.
+at::TypePtr resolveTypeNameGeneral(
```
Those functions are for mobile only. Could we rename them to *Mobile?

cccclai replied:
Address in latest commit.

iseeyuan commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -0,0 +1,107 @@
+  return ivalues;
+}
+
+bool check_zip_file(std::shared_ptr<ReadAdapterInterface> rai) {
```
Is it necessary to do this check? I don't think we are doing this for mobile load. Also is it copied from serialization/import.cpp? If so, would be good to share the code and comments.

cccclai replied:
Move the one in `serialization/import.cpp` to `serialization/import_read`, and reuse it here.

iseeyuan reviewed (approved):
LGTM in general. Left two inline comments on naming and code sharing.

Commit:
Update on "[PyTorch][Edge] Add api to get bytecode model version"


Add an api `_get_bytecode_version` to get version number given a bytecode model in both cxx and python, and the input can be both from file path and buffer.
## Test
CI (new added unit test will run as part of `pytorch_core-buck`)

1. run test_lite_interpreter.cpp
2. `python test/mobile/test_bytecode.py`



Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27961417/)!

Differential Revision: [D27961417](https://our.internmc.facebook.com/intern/diff/D27961417)

[ghstack-poisoned]
